### ISSUE TITLE
Grid with no columns contains no header bar

### ISSUE THREAD

**Issue by mvysny** [vaadin/vaadin-grid-flow#912]
This code will make the Grid to only have one header row:
```
        final Grid<Person> grid = new Grid<>(Person.class, false);
        final HeaderRow filterBar = grid.appendHeaderRow();
        final Grid.Column<Person> col = grid.addColumn("firstName");
        filterBar.getCell(col).setComponent(new TextField());
        add(grid);
```

Simply swapping two rows around will cause the Grid to have two header rows:
```
        final Grid<Person> grid = new Grid<>(Person.class, false);
        final Grid.Column<Person> col = grid.addColumn("firstName");
        final HeaderRow filterBar = grid.appendHeaderRow();
        filterBar.getCell(col).setComponent(new TextField());
        add(grid);
```

I believe this behaviour is inconsistent and both code snippets should cause the Grid to have two header rows.

**Comment by mvysny** [comment]
PR: https://github.com/vaadin/vaadin-grid-flow/pull/985

**Comment by pekam** [comment]
The inconsistency is less disturbing when you realize that `addColumn(String)` overload automatically sets a header for the column based on the property name.

So, to focus on the relevant parts and make them visible, the first case is:
1. `grid.appendHeaderRow()`
2. `column.setHeader()`

And the second case is:
1. `column.setHeader()`
2. `grid.appendHeaderRow()`

The API design problem here is, that it's not obvious how `setHeader` should work with the header row APIs. The way it has been implemented is:
- If there is no header row yet, `setHeader` will create the first row automatically
- If there already is a header row, `setHeader` will use that one instead of creating another one

I think that pretty much the best way these two kinds of APIs can work together.

**Comment by pekam** [comment]
> PR: #985

This would be a breaking change. Also, it's not always desirable to have headers at all, so always adding a header row doesn't sound like a good idea.

**Comment by mvysny** [comment]
My idea was that removing the current inconsistency in the API is more important than backward incompatibility:

1. new Grid() doesn't add header
2. new Grid(Class) does add header
3. new Grid(Class, true) does add header
4. new Grid(Class, false) does not add the header

However, this enhancement is blocked by https://github.com/vaadin/vaadin-grid-flow/issues/480 - it's not possible to remove a header and therefore it would not be possible to have a Grid without any headers.

**Comment by pekam** [comment]
I'm not sure. Adding the headers by default is part of the magic of the bean grid feature, ie. auto-generating columns from bean type's properties. I see how this is confusing from your perspective, but it's kind of a corner case where this occurs:
- Using `Grid(Class, false)` to construct a bean grid but disabling the automatic column generation
- Then adding header row at this point
- Using `addColumn(String)`, the overload which uses the bean grid feature to generate columns for bean properties including headers

So I wouldn't consider this as a very serious inconsistency, and like I explained earlier, I don't see how `setHeader` and the header row APIs could work differently.

You're right that the header's should be removable before adding the first header row by default could be an option, but even in that case, it shouldn't be done just to fix this inconsistency. The question to ask is: Would it make sense that a header row is added by default? My personal opinion is, that it's better to not add it until the user explicitly sets the headers or adds the header rows.

**Comment by mvysny** [comment]
Thank you :) I believe that Grid is most widely used for display of tabular lazy-loaded data, and in all of such cases the Grid has a header. Check out examples at https://vaadin.com/components/vaadin-grid/java-examples - all examples have the header bar visible. The header bar is used to inform the user about the name of the values in column, to sort, to unselect all (in case of multi-select), possibly as a column visibility picker in the future...

I believe that the Grid doesn't have a header only in highly specific cases, for example when it's being used as a lazy view scroller (a Grid with 1 column).

**Comment by pekam** [comment]
I agree, in most cases the grid has headers. But the user still has to define the header content (except when using bean grid). In most cases, the user calls `setHeader`, which creates the first header row automatically if needed. So in the common case, it doesn't make any difference if the first header row would have been there from the beginning, or created on demand when `setHeader` was called.

If you don't set any headers, would you expect to still see a header row with empty cells? I'd not, but this is my personal opinion, and not a strong one.

Anyway, we have deviated quite far from the topic of this ticket. :) Maybe it would make sense to create a separate issue about your suggestion, and link to this discussion? That would make it easier to discover.

**Comment by mvysny** [comment]
Hmmm... Actually I'm thinking of renaming this issue. The original title was poorly chosen, since an empty Grid with no columns is not really a viable use-case :-) My original intent was for the Grid to always have one header pre-created, regardless of which constructor you call. If you agree, I can rename the title of this issue; if not, I'll create a new issue. Please let me know.

**Comment by pekam** [comment]
Sorry, my bad; the issue title actually describes this very well. I thought this was just about the inconsistency depending on the call order of `appendHeaderRow` and `setHeader`. No changes needed. :)

**Comment by pekam** [comment]
Oops, I thought this was just about the inconsistency depending on the call order of `appendHeaderRow` and `setHeader`, but the title was actually about missing header bar. I agree, the title could be clarified as you suggested. :)

**Comment by pekam** [comment]
Regarding your PR, I'm not the product owner, but unfortunately I see it unlikely that in the near future we would
1. Prioritize #480 and get it fixed
2. Decide to change the default behavior of Grid to automatically add the first header row as suggested

**Comment by mvysny** [comment]
I understand, no problem :) It's good that the issue is known; if it attracts enough thumbs-ups, maybe it will be re-prioritized.
